src/gfx_layout.cpp

```diff
@@ -245,33 +246,32 @@ Point Layouter::GetCharPosition(std::string_view::const_iterator ch) const
 	}
 
 	/* Initial position, returned if character not found. */
-	static const std::vector<Point> zero = { {0, 0} };
-	auto position = zero.begin();
+	const Point initial_position = {_current_text_dir == TD_LTR ? 0 : line->GetWidth(), 0};
+	const Point *position = &initial_position;
 
 	/* We couldn't find the code point index. */
 	if (str != ch) return *position;
 
 	/* Valid character. */
 
 	/* Scan all runs until we've found our code point index. */
+	size_t best_index = SIZE_MAX;
 	for (int run_index = 0; run_index < line->CountRuns(); run_index++) {
 		const ParagraphLayouter::VisualRun &run = line->GetVisualRun(run_index);
 		const auto &positions = run.GetPositions();
 		const auto &charmap = run.GetGlyphToCharMap();
 
-		/* Run starts after our character, use the last found position. */
-		if (static_cast<size_t>(charmap.front()) > index) return *position;
+		auto itp = positions.begin();
+		for (auto it = charmap.begin(); it != charmap.end(); ++it, ++itp) {
+			const size_t cur_index = static_cast<size_t>(*it);
+			/* Found exact match? */
+			if (cur_index == index) return *itp;
 
-		position = positions.begin();
-		for (auto it = charmap.begin(); it != charmap.end(); /* nothing */) {
-			/* Plain honest-to-$deity match. */
-			if (static_cast<size_t>(*it) == index) return *position;
-			++it;
-			if (it == charmap.end()) break;
-
-			/* We just passed our character, it's probably a ligature, use the last found position. */
-			if (static_cast<size_t>(*it) > index) return *position;
-			++position;
+			/* No exact match, but maybe this is the best match. */
+			if (cur_index < index && (best_index < cur_index || best_index == SIZE_MAX)) {
+				best_index = cur_index;
+				position = &*itp;
+			}
```

Comment:
It's not immediately clear to me what is considered the best match. Is it the position just before the selected ligature/character from the original string?

Reply:
Added a code comment to hopefully explain this better